Error while checking build/O2/fullCI_slc9 for 2c6cb884f9dd6793c3b61067e7a456fc04bfc0ed at 2025-12-15 12:14:
```
## sw/BUILD/O2Physics-latest/log
/sw/slc9_x86-64/O2/14922-slc9_x86-64-local1/include/CommonUtils/EnumFlags.h:146:60: error: integer value 32 is outside the valid range of values [0, 31] for the enumeration type 'RCTSelectionFlags' [-Wenum-constexpr-conversion]
/sw/slc9_x86-64/O2/14922-slc9_x86-64-local1/include/CommonUtils/EnumFlags.h:115:31: error: integer value 32 is outside the valid range of values [0, 31] for the enumeration type 'RCTSelectionFlags' [-Wenum-constexpr-conversion]
/sw/slc9_x86-64/O2/14922-slc9_x86-64-local1/include/CommonUtils/EnumFlags.h:75:51: error: integer value 32 is outside the valid range of values [0, 31] for the enumeration type 'RCTSelectionFlags' [-Wenum-constexpr-conversion]
/sw/slc9_x86-64/O2/14922-slc9_x86-64-local1/include/CommonUtils/EnumFlags.h:120:32: error: integer value 32 is outside the valid range of values [0, 31] for the enumeration type 'RCTSelectionFlags' [-Wenum-constexpr-conversion]
/sw/slc9_x86-64/O2/14922-slc9_x86-64-local1/include/CommonUtils/EnumFlags.h:121:29: error: integer value 32 is outside the valid range of values [0, 31] for the enumeration type 'RCTSelectionFlags' [-Wenum-constexpr-conversion]
/sw/slc9_x86-64/O2/14922-slc9_x86-64-local1/include/CommonUtils/EnumFlags.h:124:29: error: integer value 32 is outside the valid range of values [0, 31] for the enumeration type 'RCTSelectionFlags' [-Wenum-constexpr-conversion]
/sw/slc9_x86-64/O2/14922-slc9_x86-64-local1/include/CommonUtils/EnumFlags.h:146:60: error: integer value 33 is outside the valid range of values [0, 31] for the enumeration type 'RCTSelectionFlags' [-Wenum-constexpr-conversion]
/sw/slc9_x86-64/O2/14922-slc9_x86-64-local1/include/CommonUtils/EnumFlags.h:115:31: error: integer value 33 is outside the valid range of values [0, 31] for the enumeration type 'RCTSelectionFlags' [-Wenum-constexpr-conversion]
/sw/slc9_x86-64/O2/14922-slc9_x86-64-local1/include/CommonUtils/EnumFlags.h:75:51: error: integer value 33 is outside the valid range of values [0, 31] for the enumeration type 'RCTSelectionFlags' [-Wenum-constexpr-conversion]
/sw/slc9_x86-64/O2/14922-slc9_x86-64-local1/include/CommonUtils/EnumFlags.h:120:32: error: integer value 33 is outside the valid range of values [0, 31] for the enumeration type 'RCTSelectionFlags' [-Wenum-constexpr-conversion]
/sw/slc9_x86-64/O2/14922-slc9_x86-64-local1/include/CommonUtils/EnumFlags.h:121:29: error: integer value 33 is outside the valid range of values [0, 31] for the enumeration type 'RCTSelectionFlags' [-Wenum-constexpr-conversion]
/sw/slc9_x86-64/O2/14922-slc9_x86-64-local1/include/CommonUtils/EnumFlags.h:124:29: error: integer value 33 is outside the valid range of values [0, 31] for the enumeration type 'RCTSelectionFlags' [-Wenum-constexpr-conversion]
/sw/slc9_x86-64/O2/14922-slc9_x86-64-local1/include/CommonUtils/EnumFlags.h:146:60: error: integer value 34 is outside the valid range of values [0, 31] for the enumeration type 'RCTSelectionFlags' [-Wenum-constexpr-conversion]
/sw/slc9_x86-64/O2/14922-slc9_x86-64-local1/include/CommonUtils/EnumFlags.h:115:31: error: integer value 34 is outside the valid range of values [0, 31] for the enumeration type 'RCTSelectionFlags' [-Wenum-constexpr-conversion]
/sw/slc9_x86-64/O2/14922-slc9_x86-64-local1/include/CommonUtils/EnumFlags.h:75:51: error: integer value 34 is outside the valid range of values [0, 31] for the enumeration type 'RCTSelectionFlags' [-Wenum-constexpr-conversion]
/sw/slc9_x86-64/O2/14922-slc9_x86-64-local1/include/CommonUtils/EnumFlags.h:120:32: error: integer value 34 is outside the valid range of values [0, 31] for the enumeration type 'RCTSelectionFlags' [-Wenum-constexpr-conversion]
/sw/slc9_x86-64/O2/14922-slc9_x86-64-local1/include/CommonUtils/EnumFlags.h:121:29: error: integer value 34 is outside the valid range of values [0, 31] for the enumeration type 'RCTSelectionFlags' [-Wenum-constexpr-conversion]
/sw/slc9_x86-64/O2/14922-slc9_x86-64-local1/include/CommonUtils/EnumFlags.h:124:29: error: integer value 34 is outside the valid range of values [0, 31] for the enumeration type 'RCTSelectionFlags' [-Wenum-constexpr-conversion]
/sw/slc9_x86-64/O2/14922-slc9_x86-64-local1/include/CommonUtils/EnumFlags.h:146:60: error: integer value 35 is outside the valid range of values [0, 31] for the enumeration type 'RCTSelectionFlags' [-Wenum-constexpr-conversion]
Error: /sw/slc9_x86-64/ROOT/v6-36-04-alice3-2/bin/rootcling: compilation failure (/sw/BUILD/d8598455b316f9cd91319111a114b2ce1b632767/O2Physics/Common/CCDB/G__O2PhysicsAnalysisCCDB9a856a347a_dictUmbrella.h)
ninja: build stopped: subcommand failed.
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/14922/2c6cb884f9dd6793c3b61067e7a456fc04bfc0ed/build_O2_fullCI_slc9/pretty.html).

--- a/Common/Utils/include/CommonUtils/EnumFlags.h
+++ b/Common/Utils/include/CommonUtils/EnumFlags.h
@@ -54,10 +54,12 @@ concept EnumFlagHelper = requires {
 // functions and also check via concepts expected properties of the enum.
 // This is very much inspired by much more extensive libraries like magic_enum.
 // Inspiration by its c++20 version (https://github.com/fix8mt/conjure_enum).
+// NOTE: Cannot detect if bit values past the underlying type are defined.
 template <EnumFlagHelper E>
 struct FlagsHelper final {
   using U = std::underlying_type_t<E>;
   using UMax = uint64_t; // max represetable type
+  static_assert(std::numeric_limits<U>::digits <= std::numeric_limits<UMax>::digits, "Underlying type has more digits than max supported digits");
 
   static constexpr bool isScoped() noexcept
   {
@@ -108,7 +110,7 @@ struct FlagsHelper final {
   static constexpr size_t MaxUnderScan{std::numeric_limits<U>::digits}; // Maximum digits the underlying type has
   static constexpr size_t MaxScan{MaxUnderScan + MarginScan};
 
-  // Checks if a given 'localation' contains an enum.
+  // Checks if a given 'location' contains an enum.
   template <E e>
   static constexpr bool isValid() noexcept
   {
@@ -128,14 +130,14 @@ struct FlagsHelper final {
       // check if this is an anonymous enum
       return true;
     }
-    return false;
-#else
+#elif __GNUC__
     else if constexpr (tpeek_v<e>[tp + getSpec<SVal::Start, SType::Enum_t>().size()] != '(' && tpeek_v<e>.find_first_of(getSpec<SVal::End, SType::Enum_t>(), tp + getSpec<SVal::Start, SType::Enum_t>().size()) != std::string_view::npos) {
       return true;
-    } else {
-      return false;
     }
+#else
+#error Unsupported compiler
 #endif
+    return false;
   }
 
   // Extract which values are present in the enum by checking all values in
@@ -161,7 +163,7 @@ struct FlagsHelper final {
   static constexpr auto Max_v{Values.back()};                                                          // Enum last entry
   static constexpr auto Min_u_v{static_cast<size_t>(Min_v)};                                           // Enum first entry as size_t
   static constexpr auto Max_u_v{static_cast<size_t>(Max_v)};                                           // Enum last entry as size_t
-  static_assert(Max_u_v < std::numeric_limits<U>::digits, "Max Bit is beyond allow range defered from underlying type");
+  static_assert(Max_u_v < std::numeric_limits<U>::digits, "Max Bit is beyond allow range deferred from underlying type");
   static constexpr bool isContinuous() noexcept { return (Max_u_v - Min_u_v + 1) == count(); } // Is the enum continuous
   static constexpr UMax makeMaxRep(size_t min, size_t max)
   {
@@ -258,7 +260,7 @@ struct FlagsHelper final {
   static constexpr std::optional<E> fromString(std::string_view str) noexcept
   {
     for (size_t i{0}; i < count(); ++i) {
-      if (Names[i] == str || NamesScoped[i] == str) {
+      if (isIEqual(Names[i], str) || isIEqual(NamesScoped[i], str)) {
         return Values[i];
       }
     }
@@ -277,7 +279,7 @@ struct FlagsHelper final {
     return toLower(a) == toLower(b);
   }
 
-  // Case-insensitive comparision for string_view.
+  // Case-insensitive comparison for string_view.
   static constexpr bool isIEqual(std::string_view s1, std::string_view s2) noexcept
   {
     if (s1.size() != s2.size()) {
@@ -294,7 +296,7 @@ struct FlagsHelper final {
   static constexpr std::string_view None{"none"};
   static constexpr bool hasNone() noexcept
   {
-    // check that enum does not contain memeber named 'none'
+    // check that enum does not contain member named 'none'
     for (size_t i{0}; i < count(); ++i) {
       if (isIEqual(Names[i], None)) {
         return true;
@@ -306,7 +308,7 @@ struct FlagsHelper final {
   static constexpr std::string_view All{"all"};
   static constexpr bool hasAll() noexcept
   {
-    // check that enum does not contain memeber named 'all'
+    // check that enum does not contain member named 'all'
     for (size_t i{0}; i < count(); ++i) {
       if (isIEqual(Names[i], All)) {
         return true;
@@ -332,7 +334,7 @@ concept EnumFlag = requires {
 };
 
 /**
- * \brief Classs to aggregate and manage enum-based on-off flags.
+ * \brief Class to aggregate and manage enum-based on-off flags.
  *
  * This class manages flags as bits in the underlying type of an enum (upto 64 bits), allowing
  * manipulation via enum member names. It supports operations akin to std::bitset
@@ -355,6 +357,7 @@ concept EnumFlag = requires {
 template <EnumFlag E>
 class EnumFlags
 {
+  static constexpr int DefaultBase{2};
   using H = details::enum_flags::FlagsHelper<E>;
   using U = std::underlying_type_t<E>;
   U mBits{0};
@@ -388,9 +391,10 @@ class EnumFlags
     std::for_each(flags.begin(), flags.end(), [this](const E f) noexcept { mBits |= to_bit(f); });
   }
   // Init from a string.
-  EnumFlags(const std::string& str)
+  //
+  explicit EnumFlags(const std::string& str, int base = DefaultBase)
   {
-    set(str);
+    set(str, base);
   }
   // Destructor.
   constexpr ~EnumFlags() = default;
@@ -413,14 +417,14 @@ class EnumFlags
   // Sets flags from a string representation.
   // This can be either from a number representation (binary or digits) or
   // a concatenation of the enums members name e.g., 'Enum1|Enum2|...'
-  void set(const std::string& s = "", int base = 2)
+  void set(const std::string& s, int base = DefaultBase)
   {
-    // on throw restore previous state and rethrow
-    const U prev = mBits;
-    reset();
     if (s.empty()) { // no-op
       return;
     }
+    // on throw restore previous state and rethrow
+    const U prev = mBits;
+    reset();
     try {
       setImpl(s, base);
     } catch (const std::exception& e) {
@@ -441,39 +445,42 @@ class EnumFlags
   }
 
   // Resets a specific flag.
-  template <typename T>
-    requires std::is_same_v<T, E>
+  template <std::same_as<E> T>
   constexpr void reset(T t)
   {
     mBits &= ~to_bit(t);
   }
 
   // Tests if a specific flag is set.
-  template <typename T>
-    requires std::is_same_v<T, E>
+  template <std::same_as<E> T>
   [[nodiscard]] constexpr bool test(T t) const noexcept
   {
     return (mBits & to_bit(t)) != None;
   }
 
   // Tests if all specified flags are set.
-  template <typename... Ts>
+  template <std::same_as<E>... Ts>
   [[nodiscard]] constexpr bool test(Ts... flags) const noexcept
   {
     return ((test(flags) && ...));
   }
 
   // Sets a specific flag.
-  template <typename T>
-    requires std::is_same_v<T, E>
+  template <std::same_as<E> T>
   constexpr void set(T t) noexcept
   {
     mBits |= to_bit(t);
   }
 
+  // Sets multiple specific flags.
+  template <std::same_as<E>... Ts>
+  constexpr void set(Ts... flags) noexcept
+  {
+    (set(flags), ...);
+  }
+
   // Toggles a specific flag.
-  template <typename T>
-    requires std::is_same_v<T, E>
+  template <std::same_as<E> T>
   constexpr void toggle(T t) noexcept
   {
     mBits ^= to_bit(t);
@@ -538,8 +545,7 @@ class EnumFlags
   }
 
   // Check if given flag is set.
-  template <typename T>
-    requires std::is_same_v<T, E>
+  template <std::same_as<E> T>
   [[nodiscard]] constexpr bool operator[](const T t) const noexcept
   {
     return test(t);
@@ -564,8 +570,7 @@ class EnumFlags
   constexpr EnumFlags& operator=(EnumFlags&& o) = default;
 
   // Performs a bitwise OR with a flag.
-  template <typename T>
-    requires std::is_same_v<T, E>
+  template <std::same_as<E> T>
   constexpr EnumFlags& operator|=(T t) noexcept
   {
     mBits |= to_bit(t);
@@ -573,8 +578,7 @@ class EnumFlags
   }
 
   // Performs a bitwise AND with a flag.
-  template <typename T>
-    requires std::is_same_v<T, E>
+  template <std::same_as<E> T>
   constexpr EnumFlags& operator&=(T t) noexcept
   {
     mBits &= to_bit(t);
@@ -582,8 +586,7 @@ class EnumFlags
   }
 
   // Returns a flag set with a bitwise AND.
-  template <typename T>
-    requires std::is_same_v<T, E>
+  template <std::same_as<E> T>
   constexpr EnumFlags operator&(T t) const noexcept
   {
     return EnumFlags(mBits & to_bit(t));
@@ -685,32 +688,89 @@ class EnumFlags
   // Set implementation, bits was zeroed before.
   void setImpl(const std::string& s, int base = 2)
   {
+    // Helper to check if character is valid for given base
+    auto isValidForBase = [](unsigned char c, int base) -> bool {
+      if (base == 2) {
+        return c == '0' || c == '1';
+      }
+      if (base == 10) {
+        return std::isdigit(c);
+      }
+      if (base == 16) {
+        return std::isdigit(c) || (c >= 'a' && c <= 'f') || (c >= 'A' && c <= 'F');
+      }
+      return false;
+    };
+
+    // hex
+    if (base == 16) {
+      std::string_view hex_str{s};
+      // Strip optional 0x or 0X prefix
+      if (s.size() >= 2 && s[0] == '0' && (s[1] == 'x' || s[1] == 'X')) {
+        hex_str.remove_prefix(2);
+      }
+      if (hex_str.empty()) {
+        throw std::invalid_argument("Empty hexadecimal string.");
+      }
+      if (!std::all_of(hex_str.begin(), hex_str.end(), [&](unsigned char c) { return isValidForBase(c, 16); })) {
+        throw std::invalid_argument("Invalid hexadecimal string.");
+      }
+      typename H::UMax v = std::stoul(std::string(hex_str), nullptr, 16);
+      if (v > H::MaxRep) {
+        throw std::out_of_range("Value exceeds enum range.");
+      }
+      mBits = static_cast<U>(v);
+      return;
+    }
+
+    // decimal and binary
     if (std::all_of(s.begin(), s.end(), [](unsigned char c) { return std::isdigit(c); })) {
-      if (base == 2) { // check of only 0 and 1 in string
-        if (!std::all_of(s.begin(), s.end(), [](char c) { return c == '0' || c == '1'; })) {
+      if (base == 2) {
+        // Binary: check only 0 and 1
+        if (!std::all_of(s.begin(), s.end(), [&](unsigned char c) { return isValidForBase(c, 2); })) {
           throw std::invalid_argument("Invalid binary string.");
         }
       }
-      typename H::UMax v = std::stoul(s, nullptr, base);
+      typename H::UMax v = std::stoul(std::string(s), nullptr, base);
       if (v > H::MaxRep) {
-        throw std::out_of_range("Values exceeds enum range.");
+        throw std::out_of_range("Value exceeds enum range.");
       }
       mBits = static_cast<U>(v);
-    } else if (std::all_of(s.begin(), s.end(), [](unsigned char c) { return std::isalnum(c) != 0 || c == '|' || c == ' ' || c == ':' || c == ',' || c == ';'; })) {
+    }
+    // enum name strings
+    else if (std::all_of(s.begin(), s.end(), [](unsigned char c) { return std::isalnum(c) != 0 || c == '|' || c == ' ' || c == ':' || c == ',' || c == ';'; })) {
       std::string cs{s};
       std::transform(cs.begin(), cs.end(), cs.begin(), [](unsigned char c) { return std::tolower(c); });
+
       if (cs == H::All) {
         mBits = All;
       } else if (cs == H::None) {
         mBits = None;
       } else {
-        // accept as delimiter ' ', '|', ';', ','
+        // Detect delimiter and ensure only one type is used
         char token = ' ';
-        std::string::size_type pos = s.find_first_of(",|;");
-        if (pos != std::string::npos) {
-          token = s[pos];
+        size_t pipePos = s.find('|');
+        size_t commaPos = s.find(',');
+        size_t semiPos = s.find(';');
+
+        // Count how many different delimiters exist
+        int delimiterCount = (pipePos != std::string_view::npos ? 1 : 0) +
+                             (commaPos != std::string_view::npos ? 1 : 0) +
+                             (semiPos != std::string_view::npos ? 1 : 0);
+
+        if (delimiterCount > 1) {
+          throw std::invalid_argument("Mixed delimiters not allowed!");
         }
-        for (const auto& tok : Str::tokenize(s, token)) {
+
+        if (pipePos != std::string_view::npos) {
+          token = '|';
+        } else if (commaPos != std::string_view::npos) {
+          token = ',';
+        } else if (semiPos != std::string_view::npos) {
+          token = ';';
+        }
+
+        for (const auto& tok : Str::tokenize(std::string(s), token)) {
           if (auto e = H::fromString(tok)) {
             mBits |= to_bit(*e);
           } else {

--- a/Common/Utils/test/testEnumFlags.cxx
+++ b/Common/Utils/test/testEnumFlags.cxx
@@ -74,11 +74,22 @@ BOOST_AUTO_TEST_CASE(Flags_test)
   multipleFlags.reset();
   BOOST_TEST(!multipleFlags.any());
 
+  // Test multiset
+  multipleFlags.reset();
+  multipleFlags.set(TestEnum::Bit2, TestEnum::Bit4);
+  BOOST_TEST(!multipleFlags.test(TestEnum::Bit1));
+  BOOST_TEST(multipleFlags.test(TestEnum::Bit2));
+  BOOST_TEST(!multipleFlags.test(TestEnum::Bit3));
+  BOOST_TEST(multipleFlags.test(TestEnum::Bit4));
+  BOOST_TEST(!multipleFlags.test(TestEnum::Bit5VeryLongName));
+
   // Test operator|
   EFlags combinedFlags = flag1 | EFlags(TestEnum::Bit2);
   BOOST_TEST(combinedFlags.test(TestEnum::Bit1));
   BOOST_TEST(combinedFlags.test(TestEnum::Bit2));
   BOOST_TEST(!combinedFlags.test(TestEnum::Bit3));
+  combinedFlags |= TestEnum::Bit5VeryLongName;
+  BOOST_TEST(combinedFlags.test(TestEnum::Bit5VeryLongName));
 
   // Test operator[]
   BOOST_TEST(combinedFlags[TestEnum::Bit1]);
@@ -304,5 +315,339 @@ BOOST_AUTO_TEST_CASE(Flags_test)
     o2::utils::EnumFlags<TestEnumLong> test("Bit32 Bit34");
     BOOST_CHECK(test.test(TestEnumLong::Bit32, TestEnumLong::Bit34));
     BOOST_CHECK(!test.test(TestEnumLong::Bit1, TestEnumLong::Bit23));
+  }
+}
+
+BOOST_AUTO_TEST_CASE(Flags_case_insensitive_test)
+{
+  using EFlags = o2::utils::EnumFlags<TestEnum>;
+
+  // Test case-insensitive flag names
+  {
+    EFlags flags("bit1"); // lowercase
+    BOOST_CHECK(flags.test(TestEnum::Bit1));
+    BOOST_CHECK(!flags.test(TestEnum::Bit2));
+  }
+
+  {
+    EFlags flags("BIT2"); // uppercase
+    BOOST_CHECK(flags.test(TestEnum::Bit2));
+    BOOST_CHECK(!flags.test(TestEnum::Bit1));
+  }
+
+  {
+    EFlags flags("BiT3"); // mixed case
+    BOOST_CHECK(flags.test(TestEnum::Bit3));
+  }
+
+  {
+    EFlags flags("bit1|BIT2|BiT3"); // mixed case with delimiter
+    BOOST_CHECK(flags.test(TestEnum::Bit1));
+    BOOST_CHECK(flags.test(TestEnum::Bit2));
+    BOOST_CHECK(flags.test(TestEnum::Bit3));
+  }
+
+  // Test special keywords case-insensitive
+  {
+    EFlags flags("ALL");
+    BOOST_CHECK(flags.all());
+  }
+
+  {
+    EFlags flags("None");
+    BOOST_CHECK(!flags.any());
+  }
+}
+
+BOOST_AUTO_TEST_CASE(Flags_error_recovery_test)
+{
+  using EFlags = o2::utils::EnumFlags<TestEnum>;
+
+  // Test that previous state is restored on exception
+  {
+    EFlags flags({TestEnum::Bit1, TestEnum::Bit2});
+    auto previousValue = flags.value();
+
+    // Try to set with invalid string
+    BOOST_CHECK_THROW(flags.set("InvalidFlag"), std::invalid_argument);
+
+    // Verify state was restored
+    BOOST_CHECK_EQUAL(flags.value(), previousValue);
+    BOOST_CHECK(flags.test(TestEnum::Bit1));
+    BOOST_CHECK(flags.test(TestEnum::Bit2));
+  }
+
+  {
+    EFlags flags({TestEnum::Bit3, TestEnum::Bit4});
+    auto previousValue = flags.value();
+
+    // Try to set with out-of-range value
+    BOOST_CHECK_THROW(flags.set("999999", 10), std::out_of_range);
+
+    // Verify state was restored
+    BOOST_CHECK_EQUAL(flags.value(), previousValue);
+    BOOST_CHECK(flags.test(TestEnum::Bit3));
+    BOOST_CHECK(flags.test(TestEnum::Bit4));
+  }
+
+  {
+    EFlags flags(TestEnum::Bit5VeryLongName);
+    auto previousValue = flags.value();
+
+    // Try to set with invalid binary string
+    BOOST_CHECK_THROW(flags.set("10102", 2), std::invalid_argument);
+
+    // Verify state was restored
+    BOOST_CHECK_EQUAL(flags.value(), previousValue);
+    BOOST_CHECK(flags.test(TestEnum::Bit5VeryLongName));
+  }
+}
+
+BOOST_AUTO_TEST_CASE(Flags_whitespace_handling_test)
+{
+  using EFlags = o2::utils::EnumFlags<TestEnum>;
+
+  // Test leading/trailing whitespace
+  {
+    EFlags flags("  Bit1  ");
+    BOOST_CHECK(flags.test(TestEnum::Bit1));
+  }
+
+  {
+    EFlags flags("  Bit1 | Bit2  ");
+    BOOST_CHECK(flags.test(TestEnum::Bit1));
+    BOOST_CHECK(flags.test(TestEnum::Bit2));
+  }
+
+  // Test excessive whitespace between flags
+  {
+    EFlags flags("Bit1    |    Bit3");
+    BOOST_CHECK(flags.test(TestEnum::Bit1));
+    BOOST_CHECK(flags.test(TestEnum::Bit3));
+    BOOST_CHECK(!flags.test(TestEnum::Bit2));
+  }
+
+  // Test tabs and other whitespace (should work with space delimiter)
+  {
+    EFlags flags("Bit1 Bit2 Bit3");
+    BOOST_CHECK(flags.test(TestEnum::Bit1));
+    BOOST_CHECK(flags.test(TestEnum::Bit2));
+    BOOST_CHECK(flags.test(TestEnum::Bit3));
+  }
+}
+
+BOOST_AUTO_TEST_CASE(Flags_count_bits_test)
+{
+  using EFlags = o2::utils::EnumFlags<TestEnum>;
+
+  // Test counting set bits
+  {
+    EFlags flags;
+    BOOST_CHECK_EQUAL(flags.count(), 0);
+  }
+
+  {
+    EFlags flags(TestEnum::Bit1);
+    BOOST_CHECK_EQUAL(flags.count(), 1);
+  }
+
+  {
+    EFlags flags({TestEnum::Bit1, TestEnum::Bit2});
+    BOOST_CHECK_EQUAL(flags.count(), 2);
+  }
+
+  {
+    EFlags flags({TestEnum::Bit1, TestEnum::Bit2, TestEnum::Bit3, TestEnum::Bit4});
+    BOOST_CHECK_EQUAL(flags.count(), 4);
+  }
+
+  {
+    EFlags flags(EFlags::All);
+    BOOST_CHECK_EQUAL(flags.count(), 5); // TestEnum has 5 members
+  }
+
+  // Test count after operations
+  {
+    EFlags flags({TestEnum::Bit1, TestEnum::Bit2, TestEnum::Bit3});
+    BOOST_CHECK_EQUAL(flags.count(), 3);
+
+    flags.reset(TestEnum::Bit2);
+    BOOST_CHECK_EQUAL(flags.count(), 2);
+
+    flags.set(TestEnum::Bit4);
+    BOOST_CHECK_EQUAL(flags.count(), 3);
+
+    flags.toggle(TestEnum::Bit1);
+    BOOST_CHECK_EQUAL(flags.count(), 2);
+  }
+}
+
+BOOST_AUTO_TEST_CASE(Flags_mixed_delimiter_validation_test)
+{
+  using EFlags = o2::utils::EnumFlags<TestEnum>;
+
+  // Test that mixed delimiters throw an error
+  {
+    BOOST_CHECK_THROW(EFlags("Bit1|Bit2,Bit3"), std::invalid_argument);
+  }
+
+  {
+    BOOST_CHECK_THROW(EFlags("Bit1;Bit2|Bit3"), std::invalid_argument);
+  }
+
+  {
+    BOOST_CHECK_THROW(EFlags("Bit1,Bit2;Bit3"), std::invalid_argument);
+  }
+
+  {
+    BOOST_CHECK_THROW(EFlags("Bit1|Bit2,Bit3;Bit4"), std::invalid_argument);
+  }
+
+  // Test that single delimiter types work
+  {
+    EFlags flags1("Bit1|Bit2|Bit3");
+    BOOST_CHECK_EQUAL(flags1.count(), 3);
+  }
+
+  {
+    EFlags flags2("Bit1,Bit2,Bit3");
+    BOOST_CHECK_EQUAL(flags2.count(), 3);
+  }
+
+  {
+    EFlags flags3("Bit1;Bit2;Bit3");
+    BOOST_CHECK_EQUAL(flags3.count(), 3);
+  }
+}
+
+BOOST_AUTO_TEST_CASE(Flags_empty_and_edge_cases_test)
+{
+  using EFlags = o2::utils::EnumFlags<TestEnum>;
+
+  // Test empty string
+  {
+    EFlags flags({TestEnum::Bit1, TestEnum::Bit2});
+    flags.set(""); // Should be no-op
+    BOOST_CHECK(flags.test(TestEnum::Bit1));
+    BOOST_CHECK(flags.test(TestEnum::Bit2));
+  }
+
+  // Test with only whitespace
+  {
+    EFlags flags({TestEnum::Bit1});
+    flags.set("   "); // Should result in empty after tokenization
+    // Depending on implementation, this might clear or throw
+    // Adjust expectation based on actual behavior
+  }
+
+  // Test duplicate flags (should work, setting same bit twice is idempotent)
+  {
+    EFlags flags("Bit1|Bit1|Bit1");
+    BOOST_CHECK(flags.test(TestEnum::Bit1));
+    BOOST_CHECK_EQUAL(flags.count(), 1);
+  }
+
+  // Test scoped and unscoped mixed
+  {
+    EFlags flags("Bit1|TestEnum::Bit2");
+    BOOST_CHECK(flags.test(TestEnum::Bit1));
+    BOOST_CHECK(flags.test(TestEnum::Bit2));
+  }
+}
+
+BOOST_AUTO_TEST_CASE(Flags_binary_decimal_parsing_test)
+{
+  using EFlags = o2::utils::EnumFlags<TestEnum>;
+
+  // Test binary parsing
+  {
+    EFlags flags("101", 2);
+    BOOST_CHECK(flags.test(TestEnum::Bit1));  // bit 0
+    BOOST_CHECK(!flags.test(TestEnum::Bit2)); // bit 1
+    BOOST_CHECK(flags.test(TestEnum::Bit3));  // bit 2
+  }
+
+  // Test decimal parsing
+  {
+    EFlags flags("7", 10); // 7 = 0b111
+    BOOST_CHECK(flags.test(TestEnum::Bit1));
+    BOOST_CHECK(flags.test(TestEnum::Bit2));
+    BOOST_CHECK(flags.test(TestEnum::Bit3));
+    BOOST_CHECK(!flags.test(TestEnum::Bit4));
+  }
+
+  // Test hexadecimal parsing
+  {
+    EFlags flags("F", 16); // 15 = 0b1111
+    BOOST_CHECK(flags.test(TestEnum::Bit1));
+    BOOST_CHECK(flags.test(TestEnum::Bit2));
+    BOOST_CHECK(flags.test(TestEnum::Bit3));
+    BOOST_CHECK(flags.test(TestEnum::Bit4));
+    BOOST_CHECK(!flags.test(TestEnum::Bit5VeryLongName));
+  }
+
+  // Test hexadecimal with 0x prefix
+  {
+    EFlags flags("0xA", 16); // 10 = 0b1010
+    BOOST_CHECK(!flags.test(TestEnum::Bit1));
+    BOOST_CHECK(flags.test(TestEnum::Bit2));
+    BOOST_CHECK(!flags.test(TestEnum::Bit3));
+    BOOST_CHECK(flags.test(TestEnum::Bit4));
+  }
+
+  // Test hexadecimal with 0X prefix (uppercase)
+  {
+    EFlags flags("0X1F", 16); // 31 = all 5 bits
+    BOOST_CHECK(flags.all());
+  }
+
+  // Test lowercase hex digits
+  {
+    EFlags flags("0xa", 16);
+    BOOST_CHECK_EQUAL(flags.value(), 10);
+  }
+
+  // Test thros
+  {
+    BOOST_CHECK_THROW(EFlags("0xAbCd", 16), std::out_of_range);
+  }
+
+  // Test invalid binary string (contains 2)
+  {
+    BOOST_CHECK_THROW(EFlags("1012", 2), std::invalid_argument);
+  }
+
+  // Test out of range for base
+  {
+    BOOST_CHECK_THROW(EFlags("100000", 2), std::out_of_range);
+  }
+}
+
+BOOST_AUTO_TEST_CASE(Flags_operator_bool_test)
+{
+  using EFlags = o2::utils::EnumFlags<TestEnum>;
+
+  // Test explicit bool conversion
+  {
+    EFlags empty;
+    BOOST_CHECK(!static_cast<bool>(empty));
+  }
+
+  {
+    EFlags withFlag(TestEnum::Bit1);
+    BOOST_CHECK(static_cast<bool>(withFlag));
+  }
+
+  // Test in conditional
+  {
+    EFlags flags;
+    if (flags) {
+      BOOST_FAIL("Empty flags should be false");
+    }
+
+    flags.set(TestEnum::Bit1);
+    if (!flags) {
+      BOOST_FAIL("Non-empty flags should be true");
+    }
   }
 }

